cli/paas/install_client.go

```diff
@@ -21,88 +38,92 @@ type InstallClient struct {
-	// back to cli option and/or interactive entry by the user.
-	//
-	// NOTE: This is function is set here and not in the gitea
-	// definition because the function has to have access to the
-	// cluster in question, and that information is only available
-	// now, not at deployment declaration time.
-
-	domain, err := deployments.NeededOptions.GetOpt("system_domain", "")
+	var err error
+	options, err = options.Populate(kubernetes.NewCLIOptionsReader(cmd))
 	if err != nil {
 		return errors.Wrap(err, "Couldn't install carrier")
 	}
-	domain.DynDefaultFunc = func(o *kubernetes.InstallationOption) error {
-		ips := c.kubeClient.GetPlatform().ExternalIPs()
-		if len(ips) > 0 {
-			domain := fmt.Sprintf("%s.nip.io", ips[0])
-			o.Value = domain
-		}
-		// else leave invalid, to be handled by cli option
-		// reader or interactive entry
-		return nil
-	}
-
-	deployments.PopulateNeededOptions(kubernetes.NewCLIOptionsReader(cmd))
 
 	interactive, err := cmd.Flags().GetBool("interactive")
 	if err != nil {
 		return errors.Wrap(err, "Couldn't install carrier")
 	}
 
 	if interactive {
-		deployments.PopulateNeededOptions(kubernetes.NewInteractiveOptionsReader(os.Stdout, os.Stdin))
+		options, err = options.Populate(kubernetes.NewInteractiveOptionsReader(os.Stdout, os.Stdin))
+		if err != nil {
+			return errors.Wrap(err, "Couldn't install carrier")
+		}
 	} else {
-		deployments.PopulateNeededOptions(kubernetes.NewDefaultOptionsReader())
+		options, err = options.Populate(kubernetes.NewDefaultOptionsReader())
+		if err != nil {
+			return errors.Wrap(err, "Couldn't install carrier")
+		}
 	}
 
-	c.showInstallConfiguration(deployments)
+	c.showInstallConfiguration(options)
 
 	// TODO (post MVP): Run a validation phase which perform
 	// additional checks on the values. For example range limits,
 	// proper syntax of the string, etc. do it as pghase, and late
 	// to report all problems at once, instead of early and
 	// piecemal.
 
-	err = c.deploySet(deployments)
+	deployment := deploymentSet["traefik"]
+	err = deployment.Deploy(c.kubeClient, c.ui, options.ForDeployment(deployment.ID()))
+	if err != nil {
+		return err
+	}
+
+	// Try to give a nip.io domain if the user didn't specify one
+	domain, err := options.GetOpt("system_domain", "")
+	if err != nil {
+		return errors.Wrap(err, "Couldn't install carrier")
+	}
+
+	err = c.fillInMissingSystemDomain(domain)
 	if err != nil {
 		return errors.Wrap(err, "Couldn't install carrier")
 	}
+	if domain.Value.(string) == "" {
+		return errors.New("You didn't provide a system_domain and we were unable to setup a nip.io domain (couldn't find and ExternalIP)")
+	}
+
+	c.ui.Success().Msg("Created system_domain: " + domain.Value.(string))
+
+	for _, deploymentName := range []string{"quarks", "gitea", "eirini", "registry", "tekton"} {
+		deployment := deploymentSet[deploymentName]
+		options := options.ForDeployment(deployment.ID())
+		err := deployment.Deploy(c.kubeClient, c.ui, options)
```

Comment:
See above, the entire thing can be moved into `Deploy`.

Reply:
I believe this would be easier with a simple `[]deployment` instead of a map. Keep `traefik` separate, and then you can simply iterate directly over deployments here, without names and temp vars.

See the other simplification as well, noted above.

Reply:
In the `Uninstall` method below I need to handle eirini first, like I do here with traefik. Since I need to have random access to deployments of the same list, a map is more suitable.

If I switch to a slice, I would need to keep eirini separate in the `Uninstall` method and then iterate over a different slice for the rest of them (different from the one used in `Install` I mean). This could lead to Deployments added/removed from the `Install` slice but not from the `Uninstall` and vise versa.

Reply:
Yes to what going to slice entails. Danger of de-syncing the slices. True. (Although I believe it low, when the slices are defined side by side). We can keep the map. I would still get rid of the temp variables.

Reply:
The argument re slices is not fully valid, I just realized. Because right now, even with the map you are handling slices. not of the deployments, but of the names, and these can desync as well. Especially as they are quite far apart, in the install and uninstall functions. With the deployments the plan was to have them together at the top.

Regardless, let us keep the map for now, and discuss later.

Reply:
Yes I realized that too and went with your suggestion. Will push a commit shortly.

Reply:
Heh. Flip-flop :wink: